File: src/rules/dockerfile/no_dockerignore.rs
use super::{DockerfileRule, Issue, Severity, ImpactEstimate};
use crate::parser::DockerfileParser;
use crate::rules::Rule;
use std::path::Path;

pub struct NoDockerignoreRule;

impl Rule for NoDockerignoreRule {
    fn id(&self) -> &'static str { "DF003" }
    fn name(&self) -> &'static str { "No .dockerignore" }
    fn severity(&self) -> Severity { Severity::Warning }
    
    fn description(&self) -> &'static str {
        "No .dockerignore file found in the build context"
    }
    
    fn rationale(&self) -> &'static str {
        "Without a .dockerignore file, Docker copies all files in the build context, \
         including node_modules, .git, build artifacts, and other unnecessary files. \
         This bloats the image, slows down builds, and may leak sensitive files."
    }
    
    fn fix_suggestion(&self) -> Option<&'static str> {
        Some("Create a .dockerignore file with patterns like: .git, node_modules, *.log, .env")
    }
    
    fn impact(&self) -> Option<ImpactEstimate> {
        Some(ImpactEstimate {
            build_time_improvement: Some("Can reduce build context transfer by 50-90%".to_string()),
            image_size_reduction: Some("Can reduce image size by 20-80%".to_string()),
            security_improvement: Some("Prevents accidental inclusion of secrets".to_string()),
            reliability_improvement: None,
        })
    }
}

impl DockerfileRule for NoDockerignoreRule {
    fn check(&self, _parser: &DockerfileParser, context_dir: Option<&Path>) -> Vec<Issue> {
        let mut issues = Vec::new();
        
        if let Some(dir) = context_dir {
            if !crate::parser::check_dockerignore(dir) {
                issues.push(Issue {
                    rule_id: self.id().to_string(),
                    rule_name: self.name().to_string(),
                    severity: self.severity(),
                    line_number: None,
                    message: "No .dockerignore file found in the build context".to_string(),
                    fix_suggestion: self.fix_suggestion().map(|s| s.to_string()),
                    impact: self.impact(),
                });
            }
        }
        
        issues
    }
}

File: internal/rules/dockerfile/no_dockerignore.go
package dockerfile

import (
	"github.com/thirukguru/docker-review/internal/types"
)

// NoDockerignoreRule checks for .dockerignore (DF003)
type NoDockerignoreRule struct {
	types.BaseRule
}

func NewNoDockerignoreRule() *NoDockerignoreRule {
	return &NoDockerignoreRule{
		BaseRule: types.BaseRule{
			RuleID:          "DF003",
			RuleName:        "No .dockerignore",
			RuleSeverity:    types.Warning,
			RuleCategory:    "performance",
			RuleDescription: "No .dockerignore file found. Build context may include unnecessary files.",
			RuleFix:         "Create a .dockerignore file to exclude node_modules, .git, etc.",
		},
	}
}

func (r *NoDockerignoreRule) Check(ctx *types.DockerfileContext) []types.Issue {
	var issues []types.Issue

	if !ctx.HasDockerignore {
		issues = append(issues, types.Issue{
			RuleID:      r.ID(),
			Name:        r.Name(),
			Severity:    r.Severity(),
			Line:        1,
			Description: r.Description(),
			Fix:         r.Fix(),
			Category:    r.Category(),
		})
	}

	return issues
}
